### ISSUE TITLE
no-unmodified-loop-condition issue inside for loop

### ISSUE THREAD

**Issue by lcrocker** [eslint/eslint#8788]
I don't have the ability to reopen it, but I have added a complete test case that works in the online demo and illustrates that the problem is still there. Please re-open bug #8752 (and of course close this one as a duplicate).

**Comment by gyandeeps** [comment]
I investigated further it seems if the comparison has a `MemberExpression` then the error is not shown but if it has a `Identifier` or `Literal` then the problem is shown.

```js
for (let i = 1; i < x.test; i + 1) {  
}
```

**Comment by platinumazure** [comment]
I think the reason for the different behavior in MemberExpression conditions is the possibility that a getter with side effects could be involved. (Note that function calls in conditions are explicitly considered as possible modifications.)

I think that realistically, those should be extremely rare in loop conditions and we should consider enhancing the rule with a new option to treat MemberExpression conditions as having no side effects, bringing the provided case in line with expectations.

I would even favor treating this as a semver-minor bug and changing the default behavior, but if we want to change the default behavior in a semver-major release, that's fine by me too.

**Comment by mysticatea** [comment]
The rule reports unmodified *conditions*, not variables. In that case, ESLint cannot assume that the condition `i < x.test` is unmodified since it's a static analyzer.

**Comment by SpadeAceman** [comment]
From what I can tell, the original point of #8752 was (to put it in AST/ESQuery terminology) that `ForStatement.update` should never be a `BinaryExpression` like `i + 1`, which obviously has no side effects and whose result is discarded. The programmer almost certainly intended this to be an `AssignmentExpression` like `i += 1` instead. 

It seems reasonable to suggest that the presence of an expression in `ForStatement.update` with no possible side effects whose result is discarded (such as `i + 1`) should be considered a serious red flag. I'd say it should be enough to override any ambiguous conclusions resulting from the presence of a `MemberExpression` elsewhere.

(And I agree with @platinumazure that this rule should have an option to ignore the possibility of getters with side effects.)

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
